File: platform/internal/delivery/webhook_test.go
package delivery

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
)

func TestBuildWebhookPayloadFormats(t *testing.T) {
	notif := Notification{
		ID:          "n1",
		TenantID:    "t1",
		WorkspaceID: "w1",
		Type:        NotificationScoreDrop,
		Title:       "Skor Düştü",
		Body:        "marka skoru 10 puan azaldı",
		Data:        map[string]interface{}{"drop": 10},
	}

	// generic
	payload, ct, err := buildWebhookPayload(notif)
	if err != nil {
		t.Fatalf("generic payload: %v", err)
	}
	if ct != "application/json" {
		t.Errorf("generic ct = %s", ct)
	}
	var generic map[string]interface{}
	if err := json.Unmarshal(payload, &generic); err != nil {
		t.Fatalf("generic json: %v", err)
	}
	if generic["title"] != "Skor Düştü" {
		t.Errorf("generic title = %v", generic["title"])
	}

	// slack — text alanı var, custom_details yok
	payload, _, err = buildWebhookPayload(Notification{WebhookKind: WebhookKindSlack, Title: "T", Body: "B"})
	if err != nil {
		t.Fatalf("slack payload: %v", err)
	}
	if !strings.Contains(string(payload), `"text"`) {
		t.Errorf("slack payload text alanı yok: %s", payload)
	}

	// teams — MessageCard sabitleri
	payload, _, err = buildWebhookPayload(Notification{WebhookKind: WebhookKindTeams, Title: "T", Body: "B"})
	if err != nil {
		t.Fatalf("teams payload: %v", err)
	}
	if !strings.Contains(string(payload), `"MessageCard"`) {
		t.Errorf("teams payload MessageCard yok: %s", payload)
	}

	// discord — content alanı
	payload, _, err = buildWebhookPayload(Notification{WebhookKind: WebhookKindDiscord, Title: "T", Body: "B"})
	if err != nil {
		t.Fatalf("discord payload: %v", err)
	}
	if !strings.Contains(string(payload), `"content"`) {
		t.Errorf("discord payload content yok: %s", payload)
	}

	// pagerduty — score_drop severity warning
	payload, _, err = buildWebhookPayload(Notification{WebhookKind: WebhookKindPagerDuty, Type: NotificationScoreDrop, Title: "T", Body: "B"})
	if err != nil {
		t.Fatalf("pagerduty payload: %v", err)
	}
	var pd map[string]interface{}
	if err := json.Unmarshal(payload, &pd); err != nil {
		t.Fatalf("pagerduty json: %v", err)
	}
	pdPayload := pd["payload"].(map[string]interface{})
	if pdPayload["severity"] != "warning" {
		t.Errorf("pagerduty severity = %v (want warning)", pdPayload["severity"])
	}
}

func TestSendWebhookErrorHandling(t *testing.T) {
	// URL yoksa hata
	s := &service{}
	if err := s.SendWebhook(Notification{}); err == nil {
		t.Fatal("webhook URL yokken hata bekleniyor")
	}

	// non-2xx dönünce hata
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
	}))
	defer server.Close()

	if err := s.SendWebhook(Notification{WebhookURL: server.URL, WebhookKind: WebhookKindSlack, Title: "T", Body: "B"}); err == nil {
		t.Fatal("HTTP 500 dönünce hata bekleniyor")
	}
}

func TestSendWebhookSuccess(t *testing.T) {
	var mu sync.Mutex
	received := false
	contentType := ""

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		received = true
		contentType = r.Header.Get("Content-Type")
		mu.Unlock()
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	s := &service{}
	notif := Notification{
		WebhookURL:  server.URL,
		WebhookKind: WebhookKindDiscord,
		Title:       "T",
		Body:        "B",
	}
	if err := s.SendWebhook(notif); err != nil {
		t.Fatalf("webhook gönderimi: %v", err)
	}

	mu.Lock()
	defer mu.Unlock()
	if !received {
		t.Fatal("server webhook almadı")
	}
	if contentType != "application/json" {
		t.Errorf("content-type = %s", contentType)
	}
}

func TestSendNotificationWebhookChannel(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	s := &service{}
	notif := Notification{
		ID:          "n-webhook",
		Channel:     ChannelWebhook,
		WebhookURL:  server.URL,
		WebhookKind: WebhookKindGeneric,
		Title:       "T",
		Body:        "B",
	}
	if err := s.SendNotification(notif); err != nil {
		t.Fatalf("SendNotification(webhook): %v", err)
	}
}

File: platform/java/src/test/java/dev/geolens/delivery/WebhookTest.java
package dev.geolens.delivery;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/** Go {@code delivery/webhook_test.go} portu. */
class WebhookTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static DeliveryService svc() {
        return new DeliveryService(EmailConfig.mock(), null, null);
    }

    private static Notification notif(String channel, String kind, String title, String body, Map<String, Object> data) {
        return new Notification("n1", "t1", "", "w1", DeliveryConstants.NOTIFICATION_SCORE_DROP, channel,
                title, body, "", data, DeliveryConstants.DELIVERY_PENDING, null, Instant.now(), false, "", kind);
    }

    private static HttpServer server(int status, AtomicBoolean received, AtomicReference<String> contentType) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(0), 0);
        server.createContext("/", exchange -> {
            if (received != null) {
                received.set(true);
            }
            if (contentType != null) {
                contentType.set(exchange.getRequestHeaders().getFirst("Content-Type"));
            }
            exchange.sendResponseHeaders(status, -1);
            exchange.close();
        });
        server.start();
        return server;
    }

    @Test
    void buildWebhookPayloadFormats() throws Exception {
        Notification notif = notif(DeliveryConstants.CHANNEL_WEBHOOK, "", "Skor Düştü",
                "marka skoru 10 puan azaldı", Map.of("drop", 10));

        // generic
        DeliveryService.WebhookPayload payload = DeliveryService.buildWebhookPayload(notif);
        assertEquals("application/json", payload.contentType());
        Map<?, ?> generic = MAPPER.readValue(payload.body(), Map.class);
        assertEquals("Skor Düştü", generic.get("title"));

        // slack — text alanı var
        payload = DeliveryService.buildWebhookPayload(notif("", DeliveryConstants.WEBHOOK_KIND_SLACK, "T", "B", Map.of()));
        assertTrue(new String(payload.body(), StandardCharsets.UTF_8).contains("\"text\""));

        // teams — MessageCard sabitleri
        payload = DeliveryService.buildWebhookPayload(notif("", DeliveryConstants.WEBHOOK_KIND_TEAMS, "T", "B", Map.of()));
        assertTrue(new String(payload.body(), StandardCharsets.UTF_8).contains("MessageCard"));

        // discord — content alanı
        payload = DeliveryService.buildWebhookPayload(notif("", DeliveryConstants.WEBHOOK_KIND_DISCORD, "T", "B", Map.of()));
        assertTrue(new String(payload.body(), StandardCharsets.UTF_8).contains("\"content\""));

        // pagerduty — score_drop severity warning
        payload = DeliveryService.buildWebhookPayload(notif("", DeliveryConstants.WEBHOOK_KIND_PAGERDUTY, "T", "B", Map.of()));
        Map<?, ?> pd = MAPPER.readValue(payload.body(), Map.class);
        Map<?, ?> pdPayload = (Map<?, ?>) pd.get("payload");
        assertEquals("warning", pdPayload.get("severity"));
    }

    @Test
    void sendWebhookErrorHandling() throws Exception {
        DeliveryService s = svc();

        // URL yoksa hata
        assertThrows(DeliveryException.class,
                () -> s.sendWebhook(notif("", DeliveryConstants.WEBHOOK_KIND_SLACK, "T", "B", Map.of())));

        // non-2xx dönünce hata
        HttpServer server = server(500, null, null);
        try {
            Notification base = notif("", DeliveryConstants.WEBHOOK_KIND_SLACK, "T", "B", Map.of());
            Notification n = new Notification(base.id(), base.tenantId(), base.userId(), base.workspaceId(), base.type(),
                    base.channel(), base.title(), base.body(), base.htmlBody(), base.data(), base.status(), base.sentAt(),
                    base.createdAt(), base.isRead(),
                    "http://localhost:" + server.getAddress().getPort() + "/", base.webhookKind());
            assertThrows(DeliveryException.class, () -> s.sendWebhook(n));
        } finally {
            server.stop(0);
        }
    }

    @Test
    void sendWebhookSuccess() throws Exception {
        AtomicBoolean received = new AtomicBoolean(false);
        AtomicReference<String> contentType = new AtomicReference<>();
        HttpServer server = server(200, received, contentType);
        try {
            DeliveryService s = svc();
            Notification base = notif("", DeliveryConstants.WEBHOOK_KIND_DISCORD, "T", "B", Map.of());
            Notification n = new Notification(base.id(), base.tenantId(), base.userId(), base.workspaceId(), base.type(),
                    base.channel(), base.title(), base.body(), base.htmlBody(), base.data(), base.status(), base.sentAt(),
                    base.createdAt(), base.isRead(),
                    "http://localhost:" + server.getAddress().getPort() + "/", base.webhookKind());
            s.sendWebhook(n);
            assertTrue(received.get(), "server webhook almadı");
            assertEquals("application/json", contentType.get());
        } finally {
            server.stop(0);
        }
    }

    @Test
    void sendNotificationWebhookChannel() throws Exception {
        HttpServer server = server(200, null, null);
        try {
            DeliveryService s = svc();
            Notification base = notif(DeliveryConstants.CHANNEL_WEBHOOK, DeliveryConstants.WEBHOOK_KIND_GENERIC, "T", "B", Map.of());
            Notification n = new Notification(base.id(), base.tenantId(), base.userId(), base.workspaceId(), base.type(),
                    base.channel(), base.title(), base.body(), base.htmlBody(), base.data(), base.status(), base.sentAt(),
                    base.createdAt(), base.isRead(),
                    "http://localhost:" + server.getAddress().getPort() + "/", base.webhookKind());
            s.sendNotification(n);
        } finally {
            server.stop(0);
        }
    }
}